File: app/src/main/java/com/example/punchpad2/SearchUtils.java
package com.example.punchpad2;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;

public class SearchUtils {

    public enum SortOrder { NEWEST_FIRST, OLDEST_FIRST }

    public static List<Note> filterAndSort(List<Note> notes, String query, boolean favoritesOnly, SortOrder order) {
        List<Note> result = new ArrayList<>();

        for (Note note : notes) {
            if ((query == null || note.content.toLowerCase().contains(query.toLowerCase()))
                    && (!favoritesOnly || note.favorited)) {
                result.add(note);
            }
        }

        Comparator<Note> comparator = Comparator.comparingLong(note -> note.timestamp);
        if (order == SortOrder.NEWEST_FIRST) {
            comparator = comparator.reversed();
        }
        Collections.sort(result, comparator);

        return result;
    }
}

File: app/src/main/java/com/stratonotes/SearchUtils.kt
package com.stratonotes

import java.util.Collections
import java.util.Locale

object SearchUtils {
    fun filterAndSort(
        notes: List<Note>,
        query: String?,
        favoritesOnly: Boolean,
        order: SortOrder
    ): List<Note> {
        val result: MutableList<Note> = ArrayList()

        for (note in notes) {
            if ((query == null || note.content!!.lowercase(Locale.getDefault()).contains(
                    query.lowercase(
                        Locale.getDefault()
                    )
                ))
                && (!favoritesOnly || note.isFavorited)
            ) {
                result.add(note)
            }
        }

        var comparator = Comparator.comparingLong { note: Note -> note.timestamp }
        if (order == SortOrder.NEWEST_FIRST) {
            comparator = comparator.reversed()
        }
        Collections.sort(result, comparator)

        return result
    }

    enum class SortOrder {
        NEWEST_FIRST, OLDEST_FIRST
    }
}